### PR TITLE
#104: Implemented log_values for logging user defined values. 

### PR DESCRIPTION
**Pull request by trbedwards**
@robintw @jvdzwaan let me know if this is what you had in mind

### CODE REVIEW

**Comment by jvdzwaan** [comment]
Thanks @trbedwards! 

I see you are storing a list of dictionaries in the database. What I had in mind was a single dictionary the user can add values to (e.g., for this run I want to store `color=blue` and `favorite=whatever` and for the next it's `arbitrary_value=q`). But maybe there is a good reason to use a list of dicts. What was your reasoning behind that?

**Comment by robintw** [comment]
Thanks @trbedwards - it's really wonderful to see another PR so soon after the sprint (I assume you got home ok and everything? We arrived back this afternoon, exhausted!).

This looks great - although I do have the same question as above about the list of dictionaries. I also wonder whether there could be a way to change the function signature to make it as easy to use as possible. Ideally I think it'd be great if we could call it as:

```
log_values({'a':5, 'b':10})
log_values(a=5, b=10)
log_values(a=5)
```

The latter two could be done using `kwargs`, which is a `dict` anyway. Doing that while also allowing the first one should be manageable with an optional argument and doing a bit of checking of the keys in `kwargs` - what do you think?

In a wonderfully-ideal world I would love to be able to do `log_values(variable)` and have recipy automatically grab the name of the variable as well as the value. This would make it as 'effortless' as it could possibly be - so if we called `log_values(cutoff_threshold)` we'd get the same as doing `log_values({'cutoff_threshold':cutoff_threshold})` - but I think this might be going a bit far...! (If someone comes up with a magic way of doing that then it could be a future PR - don't hold up this one for my insane ideas...!)

**Comment by trbedwards** [comment]
I used a list because I was getting errors like `unhashable type 'dict'`, but I'll see if I can use a dictionary instead somehow

**Comment by trbedwards** [comment]
The latest commit should address these issues @jvdzwaan @robintw. Let me know what you think

File: recipy/log.py [hunk 1]
```diff
@@ -96,17 +96,27 @@ def log_init():
...
+    e.g,
+    >>> log_values(a=1, b=2)
+    >>> log_values({'c': 3, 'd': 4})
+    >>> log_values({'e': 5, 'f': 6}, g=7, h=8)
+    """
+
+    # create dictionary of custom values from arguments
+    custom_values = {} if custom_values is None else custom_values
+    assert isinstance(custom_values, dict), \
+        "custom_values must be a dict. type(custom_values) = %s" % custom_values
```
Comment (bilderbuchi) [inline recipy/log.py]: Shouldn't that be `"custom_values must be a dict. type(custom_values) = %s" % type(custom_values)`?

File: recipy/log.py [hunk 2]
```diff
@@ -219,6 +244,18 @@ def transform(element):
     return transform
 
 
+def add_dict(field, dict_of_values, no_duplicates=False):
```
Comment (robintw) [inline recipy/log.py]: It doesn't look like the `no_duplicates` parameter is used - is it needed here?

File: recipy/log.py [hunk 3]
```diff
@@ -67,7 +67,8 @@ def log_init():
         "date": datetime.datetime.utcnow(),
         "command_args": " ".join(cmd_args),
         "warnings": [],
-        "libraries": [get_version('recipy')]
+        "libraries": [get_version('recipy')],
+        "custom-values": {}
```
Comment (robintw) [inline recipy/log.py]: Very minor thing: other multiple-word keys here are split by `_` not `-` - could you alter here (and below) to be consistent with this?

File: recipy/log.py [hunk 4]
```diff
@@ -95,6 +96,30 @@ def log_init():
     sys.excepthook = log_exception
 
 
+def log_values(custom_values=None, **kwargs):
+    """ Log a dictionary of custom value-key pairs into the database
```
Comment (robintw) [inline recipy/log.py]: Given that we can now use keyword arguments, could you update this to remove 'a dictionary of' so it fits the more generic usage.

**Comment by robintw** [comment]
Thanks @trbedwards - that's great :-) All merged now!

**Comment by jvdzwaan** [comment]
Cool! Thank you @trbedwards !

**Pull request merged**